### PR TITLE
Refactor some e2e code

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

ref #4003

Check if the components are in ScalePhase every 1 second, because sometime scale is fast.

Optimize and refactor the func `WaitForTidbClusterReady`. Check all container image to be same as image in spec.

Refactor the func `MustWaitForComponentPhase`
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: tests/crd_test_util.go [hunk 1]
```diff
@@ -448,17 +454,20 @@ func (ctu *CrdTestUtil) tidbMembersReadyFn(tc *v1alpha1.TidbCluster) (bool, erro
 		return false, nil
 	}
 
-	c, found := getMemberContainer(ctu.kubeCli, ctu.tcStsGetter, ns, tc.Name, label.TiDBLabelVal)
-	if !found {
-		log.Logf("statefulset: %s/%s not found containers[name=tidb] or pod %s-0",
+	expectedImage := tc.TiDBImage()
+	containers, err := utilstatefulset.GetMemberContainersFromSts(ctu.kubeCli, ctu.tcStsGetter, ns, tidbSetName, v1alpha1.TiDBMemberType)
+	if err != nil {
+		log.Logf("statefulset: %s/%s not found containers[name=tikv] or pod %s-0",
```
Comment (csuzhangxc) [inline tests/crd_test_util.go:460]: `name=tidb`

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 2]
```diff
@@ -3037,7 +3013,7 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 					replicasLarge = 4
 					replicasSmall = 3
 				case v1alpha1.TiDBMemberType:
-					replicasLarge = 3
+					replicasLarge = 4
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:3016]: Do we need `replicasLarge = 5` for `v1alpha1.TiKVMemberType` as above?

File: tests/actions.go [hunk 3]
```diff
@@ -1240,414 +1235,413 @@ func (oa *OperatorActions) CheckUpgradeOrDie(ctx context.Context, info *TidbClus
...
-func (oa *OperatorActions) pdMembersReadyFn(tc *v1alpha1.TidbCluster) (bool, error) {
-	if tc.Spec.PD == nil {
-		log.Logf("no pd in tc spec, skip")
-		return true, nil
+func (oa *OperatorActions) memberCheckContextForTC(tc *v1alpha1.TidbCluster, component v1alpha1.MemberType) (*memberCheckContext, error) {
+	name := tc.GetName()
+	stsName := fmt.Sprintf("%s-%s", name, component)
+
+	var (
+		err            error
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply tests/actions.go:1323]: forgot to remove 👀

File: tests/actions.go [hunk 4]
```diff
@@ -1240,414 +1235,413 @@ func (oa *OperatorActions) CheckUpgradeOrDie(ctx context.Context, info *TidbClus
...
+
+	if ctx.skip {
+		log.Logf("component %s is not exist in spec, so skip", component)
+		return nil
+	}
+
+	// check statefulset
+	sts, err := oa.tcStsGetter.StatefulSets(ns).Get(context.TODO(), stsName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get sts %q: %s", sts, err)
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply tests/actions.go:1279]: yes, it's the same as before

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 2 above)
Comment (KanShiori) [reply tests/e2e/tidbcluster/tidbcluster.go:3016]: seems ok

File: tests/crd_test_util.go (hunk 1 above)
Comment (KanShiori) [reply tests/crd_test_util.go]: Done

File: tests/actions.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (KanShiori) [reply tests/actions.go]: Done

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/manager/member/tikv_scaler.go [hunk 5]
```diff
@@ -129,33 +129,39 @@ func (s *tikvScaler) ScaleIn(meta metav1.Object, oldSet *apps.StatefulSet, newSe
...
-			}
 
-			// endEvictLeader for TombStone stores
-			if err = endEvictLeaderbyStoreID(s.deps, tc, id); err != nil {
-				return err
+				setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
+				return nil
+			} else {
+				klog.Warningf("TiKV %s/%s store %s in status is not equal with store %s in label",
+					ns, podName, storeID, pod.Labels[label.StoreIDLabelKey])
```
Comment (KanShiori) [inline pkg/manager/member/tikv_scaler.go:161]: just add a log

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**